substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1050 @@
+					Self::sort_tickets(
+						metadata
+							.unsorted_tickets_count
+							.div_ceil(SEGMENT_MAX_SIZE * slots_left as u32),
+						epoch_tag,
+						&mut metadata,
+					);
+					TicketsMeta::<T>::set(metadata);
+				}
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Submit next epoch tickets candidates.
+		///
+		/// The number of tickets allowed to be submitted in one call is equal to the epoch length.
+		// TODO: maybe we must be more restrictive?
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::submit_tickets(tickets.len() as u32))]
+		pub fn submit_tickets(
+			origin: OriginFor<T>,
+			tickets: BoundedVec<TicketEnvelope, MaxTicketsFor<T>>,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
+
+			debug!(target: LOG_TARGET, "Received {} tickets", tickets.len());
+
+			let Some(verifier) = RingVerifierData::<T>::get().map(|v| v.into()) else {
+				warn!(target: LOG_TARGET, "Ring verifier key not initialized");
+				return Err("Ring verifier key not initialized".into())
+			};
+
+			let next_authorities = Self::next_authorities();
+
+			// Check tickets score
+			let next_config = Self::next_config().unwrap_or_else(|| Self::config());
+			// Current slot should be less than half of epoch length.
+			let epoch_length = T::EpochLength::get();
+			let ticket_threshold = sp_consensus_sassafras::ticket_id_threshold(
+				next_config.redundancy_factor,
+				epoch_length as u32,
+				next_config.attempts_number,
+				next_authorities.len() as u32,
+			);
+
+			// Get next epoch params
+			let randomness = NextRandomness::<T>::get();
+			let epoch_idx = EpochIndex::<T>::get() + 1;
+
+			let mut valid_tickets = BoundedVec::with_max_capacity();
+			for ticket in tickets {
+				debug!(target: LOG_TARGET, "Checking ring proof");
+
+				let ticket_id_input =
+					vrf::ticket_id_input(&randomness, ticket.body.attempt_idx, epoch_idx);
+				let Some(ticket_id_output) = ticket.signature.outputs.get(0) else {
+					debug!(target: LOG_TARGET, "Missing ticket vrf output from ring signature");
+					continue
+				};
+				let ticket_id = vrf::make_ticket_id(&ticket_id_input, &ticket_id_output);
+				if ticket_id >= ticket_threshold {
+					debug!(target: LOG_TARGET, "Ignoring ticket over threshold ({:032x} >= {:032x})", ticket_id, ticket_threshold);
+					continue
+				}
+
+				if TicketsData::<T>::contains_key(ticket_id) {
+					debug!(target: LOG_TARGET, "Ignoring duplicate ticket ({:032x})", ticket_id);
+					continue
+				}
+
+				let sign_data = vrf::ticket_body_sign_data(&ticket.body, ticket_id_input);
+
+				if ticket.signature.ring_vrf_verify(&sign_data, &verifier) {
+					TicketsData::<T>::set(ticket_id, Some(ticket.body));
+					valid_tickets
+						.try_push(ticket_id)
+						.expect("Input segment has same length as bounded destination vector; qed");
+				} else {
+					debug!(target: LOG_TARGET, "Proof verification failure for ticket ({:032x})", ticket_id);
```

Comment:
Maybe consider making `TicketId` be strongly typed (that is, make it `struct TicketId(u128)` instead of `type TicketId = u128`), implement `Display` for it to standardize how its printed, and then just use `{}` instead to print it?

Reply:
Yeah, this is a good idea